homeassistant/components/nissan_leaf/__init__.py

```diff
@@ -0,0 +1,514 @@
+            async_dispatcher_send(self.hass, SIGNAL_UPDATE_LEAF)
+            return climate_result.is_hvac_running == toggle
+
+        _LOGGER.debug("Climate result not returned by Nissan servers")
+        return False
+
+    async def async_get_location(self):
+        """Get location from Nissan servers."""
+        request = await self.hass.async_add_job(self.leaf.request_location)
+        for attempt in range(MAX_RESPONSE_ATTEMPTS):
+            if attempt > 0:
+                _LOGGER.debug("Location data not in yet. (%s) (%s). "
+                              "Waiting %s seconds", self.leaf.vin,
+                              attempt, PYCARWINGS2_SLEEP)
+                await asyncio.sleep(PYCARWINGS2_SLEEP)
+
+            location_status = await self.hass.async_add_job(
+                self.leaf.get_status_from_location, request
+            )
+
+            if location_status is not None:
+                _LOGGER.debug("Location_status=%s", location_status.__dict__)
+                break
+
+        return location_status
+
+    async def async_start_charging(self):
+        """Request start charging via Nissan servers."""
+        # Send the command to request charging is started to Nissan servers.
+        # If that completes OK then trigger a fresh update to pull the
+        # charging status from the car after waiting a minute for the
+        # charging request to reach the car.
+        result = await self.hass.async_add_job(self.leaf.start_charging)
+        if result:
+            _LOGGER.debug("Start charging sent, "
+                          "request updated data in 1 minute")
+            check_charge_at = utcnow() + timedelta(minutes=1)
+            self.next_update = check_charge_at
+            async_track_point_in_utc_time(
+                self.hass, self.async_update_data, check_charge_at)
+
+
+class LeafEntity(Entity):
+    """Base class for Nissan Leaf entity."""
+
+    def __init__(self, car):
+        """Store LeafDataStore upon init."""
+        self.car = car
+
+    def log_registration(self):
+        """Log registration."""
+        _LOGGER.debug(
+            "Registered %s component for VIN %s",
+            self.__class__.__name__, self.car.leaf.vin)
+
+    @property
+    def device_state_attributes(self):
+        """Return default attributes for Nissan leaf entities."""
+        return {
+            'next_update': self.car.next_update,
+            'last_attempt': self.car.last_check,
+            'updated_on': self.car.last_battery_response,
+            'update_in_progress': self.car.request_in_progress,
+            'location_updated_on': self.car.last_location_response,
+            'vin': self.car.leaf.vin,
+        }
+
+    async def async_added_to_hass(self):
+        """Register callbacks."""
+        self.log_registration()
+        async_dispatcher_connect(
+            self.car.hass, SIGNAL_UPDATE_LEAF, self._update_callback)
+
+    def _update_callback(self):
```

Comment:
Decorate this with `@callback` imported from core.py.

Reply:
Fixed by commit https://github.com/filcole/home-assistant/commit/66e976beec6a00329683a1966eeae659b0abf07d